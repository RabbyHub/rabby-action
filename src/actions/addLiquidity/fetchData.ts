import PQueue from 'p-queue';
import {
  AddLiquidityRequireData,
  FetchActionRequiredData,
  ParsedActionData,
} from '../../types';
import { waitQueueFinished } from '../../utils/waitQueueFinished';
import BigNumber from 'bignumber.js';

export const fetchDataAddLiquidity: FetchActionRequiredData<{
  receiver: string;
  exchange_rate: string;
  min_exchange_rate: string;
  max_exchange_rate: string;
  token0: any;
  token1: any;
}> = async (options, likeAction) => {
  const { actionData, apiProvider, sender, walletProvider, chainId } = options;

  const action =
    likeAction || (<ParsedActionData<'transaction'>>actionData).addLiquidity;

  if (!action || options.type === 'text') {
    return {};
  }

  const { receiver, exchange_rate, token0, token1 } = action;

  const fair_exchange_rate = new BigNumber(token0.price).div(token1.price);

  const diff = +new BigNumber(exchange_rate)
    .minus(fair_exchange_rate)
    .abs()
    .div(fair_exchange_rate)
    .times(100)
    .toFixed(4);

  const receiverInWallet = await walletProvider.hasAddress(receiver);

  const id =
    options.type === 'transaction'
      ? options.tx.to
      : options.actionData.contractId;
  if (!id) {
    return {};
  }

  const result: AddLiquidityRequireData = {
    id,
    protocol: null,
    bornAt: 0,
    rank: null,
    sender,
    receiverInWallet,
    hasInteraction: false,
    receiver,
    poolRate: exchange_rate,
    token0,
    token1,
    marketRate: fair_exchange_rate.toFixed(),
    diff,
  };
  const queue = new PQueue();

  queue.add(async () => {
    const contractInfo = await apiProvider.getContractInfo(id, chainId);
    if (!contractInfo) {
      result.rank = null;
    } else {
      result.rank = contractInfo.credit.rank_at;
      result.bornAt = contractInfo.create_at;

      result.protocol = contractInfo.protocol;
    }
  });

  queue.add(async () => {
    const hasInteraction = await apiProvider.hasInteraction(
      sender,
      chainId,
      id
    );
    result.hasInteraction = hasInteraction.has_interaction;
  });

  await waitQueueFinished(queue);

  return result;
};
