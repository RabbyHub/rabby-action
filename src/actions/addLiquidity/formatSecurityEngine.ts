import {
  AddLiquidityRequireData,
  FormatSecurityEngineContext,
} from '../../types';

export const formatSecurityEngineAddLiquidity: FormatSecurityEngineContext =
  async (options) => {
    const { actionData, requireData, chainId } = options;

    if (!actionData.addLiquidity || !chainId) {
      return {} as any;
    }

    const data = requireData as AddLiquidityRequireData;

    return {
      addLiquidity: {
        receiverMatch:
          data.receiver?.toLowerCase() === data.sender.toLowerCase(),
        diff: data.diff,
      },
    };
  };
