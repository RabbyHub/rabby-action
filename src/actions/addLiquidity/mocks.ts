export const preExecData = {
  abi: {
    func: 'mint',
    params: [
      [
        '0x18Bc5bcC660cf2B9cE3cd51a404aFe1a0cBD3C22',
        '0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913',
        10,
        -4970,
        -4960,
        2955759909,
        885199302,
        2940981110,
        880773306,
        '0xE2866F2291B5c5bafD2858B9247f93745108A523',
        1754366820,
        0,
      ],
    ],
  },
  abi_str:
    '{\n    "func": "mint",\n    "params": [\n        [\n            "0x18Bc5bcC660cf2B9cE3cd51a404aFe1a0cBD3C22",\n            "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",\n            10,\n            -4970,\n            -4960,\n            2955759909,\n            885199302,\n            2940981110,\n            880773306,\n            "0xE2866F2291B5c5bafD2858B9247f93745108A523",\n            1754366820,\n            0\n        ]\n    ]\n}',
  balance_change: {
    success: false,
    error: {
      code: 1002,
      msg: 'execution reverted: execution reverted',
    },
    send_token_list: [],
    receive_token_list: [],
    send_nft_list: [],
    receive_nft_list: [],
    usd_value_change: 0,
  },
  gas: {
    success: true,
    error: null,
    gas_used: 29320,
    gas_limit: 0,
    gas_ratio: 1.3,
  },
  is_gnosis: false,
  native_token: {
    id: 'base',
    chain: 'base',
    name: 'ETH',
    symbol: 'ETH',
    display_symbol: null,
    optimized_symbol: 'ETH',
    decimals: 18,
    logo_url:
      'https://static.debank.com/image/coin/logo_url/eth/6443cdccced33e204d90cb723c632917.png',
    protocol_id: '',
    price: 3629.55,
    price_24h_change: 0.018385589869839923,
    credit_score: 46503412.60907772,
    is_verified: true,
    is_scam: false,
    is_suspicious: null,
    is_core: true,
    is_wallet: true,
    time_at: null,
    low_credit_score: false,
    amount: 1e-5,
  },
  pre_exec_version: 'v2',
  pre_exec: {
    success: false,
    error: {
      code: 1002,
      msg: 'execution reverted: execution reverted',
    },
  },
  type_call: {
    action: 'mint',
    contract: '0x827922686190790b37229fd06084350e74485b72',
    contract_protocol_name: 'Aerodrome V3',
    contract_protocol_logo_url:
      'https://static.debank.com/image/project/logo_url/base_aerodrome/f02d753bc321dc8ba480f0424a686482.png',
  },
  trace_id: '663ae6c6c0ac4b1aa5651fadba316de2',
} as any;

export const parseTxData = {
  action: { type: 'cancel_tx' },
  contract_call: null,
  log_id: 441860918,
} as any;

export const txData = {
  chainId: 42161,
  from: '0x4686fd7c0bc5d06346d416808916ba1dbd3f2336',
  to: '0x4686fd7c0bc5d06346d416808916ba1dbd3f2336',
  value: '0x0',
  data: '0x',
  gas: '0x0',
  maxFeePerGas: '0xb71b00',
  maxPriorityFeePerGas: '0xb71b00',
  nonce: '0x4',
};
