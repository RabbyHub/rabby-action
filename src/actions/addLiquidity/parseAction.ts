import { ParseAction } from '../../types';
import { TokenItem } from '@rabby-wallet/rabby-api/dist/types';

interface AddLiquidityActionData {
  receiver: string;
  exchange_rate: string;
  min_exchange_rate: string;
  max_exchange_rate: string;
  token0: TokenItem;
  token1: TokenItem;
}

export const parseActionAddLiquidity: ParseAction = (options) => {
  const { data } = options;

  if (data?.type !== 'add_liquidity') {
    return {};
  }

  const {
    receiver,
    exchange_rate,
    min_exchange_rate,
    max_exchange_rate,
    token0,
    token1,
  } = data.data as unknown as AddLiquidityActionData;

  return {
    addLiquidity: {
      receiver,
      exchange_rate,
      min_exchange_rate,
      max_exchange_rate,
      token0,
      token1,
    },
  };
};
