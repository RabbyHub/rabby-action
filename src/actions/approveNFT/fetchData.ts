import PQueue from 'p-queue';
import { ApproveNFTRequireData, FetchActionRequiredData } from '../../types';
import { waitQueueFinished } from '../../utils/waitQueueFinished';
import { fetchHasInteraction } from '../../fetches/fetchHasInteraction';

export const fetchDataApproveNFT: FetchActionRequiredData<{
  spender: string;
}> = async (options, likeAction) => {
  const queue = new PQueue();
  const { sender, apiProvider, chainId, actionData, extraActionDataState } =
    options;
  const action = likeAction || actionData.approveNFT;

  if (!action) {
    return {};
  }

  const result: ApproveNFTRequireData = {
    isEOA: false,
    contract: null,
    riskExposure: 0,
    rank: null,
    bornAt: 0,
    protocol: null,
    isDanger: false,
    tokenBalance: '0',
    hasInteraction: null,
    extraState: {
      hasInteraction: () => undefined,
    },
  };

  queue.add(async () => {
    const contractInfo = await apiProvider.getContractInfo(
      action.spender,
      chainId
    );
    if (!contractInfo) {
      result.isEOA = true;
      result.rank = null;
    } else {
      result.rank = contractInfo.credit.rank_at;
      result.riskExposure = contractInfo.top_nft_spend_usd_value;
      result.bornAt = contractInfo.create_at;
      result.isDanger =
        contractInfo.is_danger.auto || contractInfo.is_danger.edit;
      result.protocol = contractInfo.protocol;
    }

    if (result.isEOA) {
      queue.add(async () => {
        const { desc } = await apiProvider.addrDesc(action.spender);
        result.bornAt = desc.born_at;
      });
    }
  });

  fetchHasInteraction({
    apiProvider,
    sender,
    chainId,
    spender: action.spender,
    extraActionDataState,
    queue,
    result,
  });

  await waitQueueFinished(queue);
  return result;
};
