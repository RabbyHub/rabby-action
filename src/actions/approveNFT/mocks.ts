export const preExecData = {
  abi: {
    func: 'approve',
    params: ['0x7a250d5630B4cF539739dF2C5dAcb4c659F2488D', 745],
  },
  abi_str:
    '{\n    "func": "approve",\n    "params": [\n        "0x7a250d5630B4cF539739dF2C5dAcb4c659F2488D",\n        745\n    ]\n}',
  balance_change: {
    success: true,
    error: null,
    send_token_list: [],
    receive_token_list: [],
    send_nft_list: [],
    receive_nft_list: [],
    usd_value_change: 0,
  },
  gas: {
    success: true,
    error: null,
    gas_used: 48759,
    gas_limit: 49146,
    gas_ratio: 1.3,
  },
  is_gnosis: false,
  native_token: {
    id: 'eth',
    chain: 'eth',
    name: 'ETH',
    symbol: 'ETH',
    display_symbol: null,
    optimized_symbol: 'ETH',
    decimals: 18,
    logo_url:
      'https://static.debank.com/image/coin/logo_url/eth/6443cdccced33e204d90cb723c632917.png',
    protocol_id: '',
    price: 2632.17,
    price_24h_change: 0.020513635693958674,
    credit_score: 100000000.0,
    is_verified: true,
    is_scam: false,
    is_suspicious: false,
    is_core: true,
    is_wallet: true,
    time_at: 1483200000.0,
    amount: 0.00147682999416643,
  },
  pre_exec_version: 'v2',
  pre_exec: {
    success: true,
    error: null,
  },
  type_single_nft_approval: {
    spender: '0x7a250d5630b4cf539739df2c5dacb4c659f2488d',
    spender_protocol_name: 'Uniswap V2',
    spender_protocol_logo_url:
      'https://static.debank.com/image/project/logo_url/uniswap2/87a541b3b83b041c8d12119e5a0d19f0.png',
    token_symbol: 'Metascapes #745',
    token_amount: 1,
    is_infinity: false,
    is_nft: true,
    nft: {
      id: '12ce1dbcf984dd9174065edc9531a09b',
      contract_id: '0x75d639e5e52b4ea5426f2fb46959b9c3099b729a',
      inner_id: '745',
      chain: 'eth',
      symbol: 'MTSCPS',
      name: 'Metascapes #745',
      description:
        '2555 hand-picked AI landscape images of a unique collaboration between Cath Simard, Ryan Newburn, Iurie Belegurschi and Artificial Intelligence, bridging the natural and the supernatural.\n\nhttps://metascapes.sloika.xyz/',
      content_type: 'image_url',
      content:
        'https://static.debank.com/image/eth_nft/local_url/12ce1dbcf984dd9174065edc9531a09b/47202001ef73e7586c53412c3bc72629.png',
      thumbnail_url:
        'https://static.debank.com/image/eth_nft/thumbnail_url/12ce1dbcf984dd9174065edc9531a09b/e5a971f66038f0159ac918b0faacd49e.png',
      total_supply: 1,
      attributes: [
        {
          trait_type: 'Type',
          value: 'Image',
        },
      ],
      detail_url:
        'https://opensea.io/assets/0x75d639e5e52b4ea5426f2fb46959b9c3099b729a/745',
      collection_id: 'eth:0x75d639e5e52b4ea5426f2fb46959b9c3099b729a',
      is_erc1155: false,
      is_erc721: true,
      pay_token: {
        id: 'eth',
        amount: 0.099,
        time_at: 1678369547.0,
        date_at: '2023-03-09',
      },
      usd_price: 151.69176000000002,
      collection: {
        id: 'eth:0x75d639e5e52b4ea5426f2fb46959b9c3099b729a',
        chain: 'eth',
        name: 'The Metascapes',
        description: null,
        logo_url:
          'https://static.debank.com/image/nft_collection/logo_url/the-metascapes/065c2cd129c3c0023c6cb3ea2051be4a.png',
        is_verified: true,
        credit_score: 216.306447552488,
        is_scam: false,
        is_suspicious: false,
        is_core: true,
        floor_price: 0.0243,
      },
    },
    token: {
      raw_amount: 1,
      raw_amount_hex_str: '0x1',
      is_infinity: false,
    },
  },
  trace_id: '182385710d964fdaba026f2a76a6dc36',
} as any;

export const parseTxData = {
  action: {
    type: 'approve_nft',
    data: {
      spender: '0x7a250d5630b4cf539739df2c5dacb4c659f2488d',
      nft: {
        id: '12ce1dbcf984dd9174065edc9531a09b',
        contract_id: '0x75d639e5e52b4ea5426f2fb46959b9c3099b729a',
        inner_id: '745',
        chain: 'eth',
        symbol: 'MTSCPS',
        name: 'Metascapes #745',
        description:
          '2555 hand-picked AI landscape images of a unique collaboration between Cath Simard, Ryan Newburn, Iurie Belegurschi and Artificial Intelligence, bridging the natural and the supernatural.\n\nhttps://metascapes.sloika.xyz/',
        content_type: 'image_url',
        content:
          'https://static.debank.com/image/eth_nft/local_url/12ce1dbcf984dd9174065edc9531a09b/47202001ef73e7586c53412c3bc72629.png',
        thumbnail_url:
          'https://static.debank.com/image/eth_nft/thumbnail_url/12ce1dbcf984dd9174065edc9531a09b/e5a971f66038f0159ac918b0faacd49e.png',
        total_supply: 1,
        attributes: [
          {
            trait_type: 'Type',
            value: 'Image',
          },
        ],
        detail_url:
          'https://opensea.io/assets/0x75d639e5e52b4ea5426f2fb46959b9c3099b729a/745',
        collection_id: 'eth:0x75d639e5e52b4ea5426f2fb46959b9c3099b729a',
        is_erc1155: false,
        is_erc721: true,
        pay_token: {
          id: 'eth',
          amount: 0.099,
          time_at: 1678369547.0,
          date_at: '2023-03-09',
        },
        usd_price: 151.69176000000002,
        collection: {
          id: 'eth:0x75d639e5e52b4ea5426f2fb46959b9c3099b729a',
          chain: 'eth',
          name: 'The Metascapes',
          description: null,
          logo_url:
            'https://static.debank.com/image/nft_collection/logo_url/the-metascapes/065c2cd129c3c0023c6cb3ea2051be4a.png',
          is_verified: true,
          credit_score: 216.306447552488,
          is_scam: false,
          is_suspicious: false,
          is_core: true,
          floor_price: 0.0243,
        },
      },
    },
  },
  contract_call: {
    func: 'approve',
    contract: {
      id: '0x75d639e5e52b4ea5426f2fb46959b9c3099b729a',
      protocol: null,
    },
  },
  log_id: 84523037,
} as any;

export const txData = {
  chainId: 1,
  data: '0x095ea7b30000000000000000000000007a250d5630b4cf539739df2c5dacb4c659f2488d00000000000000000000000000000000000000000000000000000000000002e9',
  from: '0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85',
  gas: '',
  maxFeePerGas: '0x4190ab00',
  maxPriorityFeePerGas: '0x4190ab00',
  nonce: '0x168f',
  to: '0x75d639e5e52b4ea5426f2fb46959b9c3099b729a',
  value: '0x0',
};
