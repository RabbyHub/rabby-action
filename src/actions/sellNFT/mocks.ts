export const parseTxData = {
  action: {
    type: 'sell_nft_order',
    data: {
      pay_nft: {
        id: '2374a8d8f35ed510c74ad7838f06a322',
        contract_id: '0xf494d8c3a421b6175cb1d89ea70715e1932637a9',
        inner_id: '1',
        chain: 'matic',
        symbol: 'PLANT',
        name: '#1',
        description: null,
        content_type: null,
        content: '',
        thumbnail_url: '',
        total_supply: 420,
        attributes: [],
        detail_url:
          'https://opensea.io/assets/0xf494d8c3a421b6175cb1d89ea70715e1932637a9/1',
        collection_id: 'matic:0xf494d8c3a421b6175cb1d89ea70715e1932637a9',
        is_erc1155: true,
        is_erc721: false,
        pay_token: null,
        collection: {
          id: 'matic:0xf494d8c3a421b6175cb1d89ea70715e1932637a9',
          chain: 'matic',
          name: 'PLANT.z - BLOOMER',
          description: null,
          logo_url: '',
          is_verified: true,
          credit_score: null,
          is_scam: true,
          is_suspicious: false,
          is_core: false,
          floor_price: 0.0,
        },
        amount: 90,
      },
      receive_token: {
        id: 'matic',
        chain: 'matic',
        name: 'MATIC',
        symbol: 'MATIC',
        display_symbol: null,
        optimized_symbol: 'MATIC',
        decimals: 18,
        logo_url:
          'https://static.debank.com/image/matic_token/logo_url/matic/6f5a6b6f0732a7a235131bd7804d357c.png',
        protocol_id: '',
        price: 0.525,
        price_24h_change: -0.06766116142781037,
        credit_score: 100000000.0,
        is_verified: true,
        is_scam: false,
        is_suspicious: false,
        is_core: true,
        is_wallet: true,
        time_at: null,
        amount: 41.454,
        raw_amount: '41454000000000000000',
      },
      receiver: '0xfac07fa759cae79cf85fa4f7c947ace0a2d5ed8c',
      takers: [],
      expire_at: 452312848583266388373324160190187140051835877600158453286538399315143056087,
    },
  },
  log_id: 6517750,
} as any;

export const txData = {
  types: {
    EIP712Domain: [
      {
        name: 'name',
        type: 'string',
      },
      {
        name: 'version',
        type: 'string',
      },
      {
        name: 'chainId',
        type: 'uint256',
      },
      {
        name: 'verifyingContract',
        type: 'address',
      },
    ],
    ERC1155SellOrder: [
      {
        name: 'maker',
        type: 'address',
      },
      {
        name: 'taker',
        type: 'address',
      },
      {
        name: 'expiry',
        type: 'uint256',
      },
      {
        name: 'nonce',
        type: 'uint256',
      },
      {
        name: 'erc20Token',
        type: 'address',
      },
      {
        name: 'erc20TokenAmount',
        type: 'uint256',
      },
      {
        name: 'fees',
        type: 'Fee[]',
      },
      {
        name: 'erc1155Token',
        type: 'address',
      },
      {
        name: 'erc1155TokenId',
        type: 'uint256',
      },
      {
        name: 'erc1155TokenAmount',
        type: 'uint128',
      },
      {
        name: 'hashNonce',
        type: 'uint256',
      },
    ],
    Fee: [
      {
        name: 'recipient',
        type: 'address',
      },
      {
        name: 'amount',
        type: 'uint256',
      },
      {
        name: 'feeData',
        type: 'bytes',
      },
    ],
  },
  primaryType: 'ERC1155SellOrder',
  domain: {
    chainId: 137,
    name: 'ElementEx',
    verifyingContract: '0xeaf5453b329eb38be159a872a6ce91c9a8fb0260',
    version: '1.0.0',
  },
  message: {
    erc1155Token: '0xf494d8c3a421b6175cb1d89ea70715e1932637a9',
    erc1155TokenAmount: '90',
    erc1155TokenId: '1',
    erc20Token: '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee',
    erc20TokenAmount: '41454000000000000000',
    expiry:
      '452312848583266388373324160190187140051835877600158453286538399315143056087',
    fees: [
      {
        amount: '846000000000000000',
        feeData: '0x',
        recipient: '0xd207842d66b715df6ea08cf52f025b9e2ed28788',
      },
    ],
    hashNonce: '0',
    maker: '0xfac07fa759cae79cf85fa4f7c947ace0a2d5ed8c',
    nonce: '1',
    taker: '0x0000000000000000000000000000000000000000',
  },
};
