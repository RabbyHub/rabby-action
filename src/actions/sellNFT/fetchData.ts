import { FetchActionRequiredData } from '../../types';
import { fetchDataCommon } from '../common/fetchData';

export const fetchDataSellNFT: FetchActionRequiredData = async (options) => {
  if (options.type !== 'typed_data') {
    return {};
  }
  if (!options.actionData.contractId || !options.actionData.sellNFT) {
    return {};
  }

  return fetchDataCommon(options, {
    receiver: options.actionData.contractId,
  });
};
