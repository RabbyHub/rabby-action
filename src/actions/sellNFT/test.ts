import {
  apiProvider,
  formatProvider,
  walletProvider,
  SENDER,
  ETH_CHAIN_ID,
  ORIGIN,
} from '../../../__mocks__';
import { parseTxData, txData } from './mocks';
import { parseActionSellNFT } from './parseAction';
import { fetchDataSellNFT } from './fetchData';
import { formatSecurityEngineSellNFT } from './formatSecurityEngine';
import { parseTypedDataAction } from '../../utils/parseTypedDataAction';
import {
  parseAction,
  fetchActionRequiredData,
  formatSecurityEngineContext,
} from '../..';

/**
/**
 * https://brain.debank.com/action-group/typed_data/a42fd2da1ff833684ed76cd90d71ee96
 * - contract_id: 0xeaf5453b329eb38be159a872a6ce91c9a8fb0260
 * id: 6635428
 */
test.each([
  [
    parseTypedDataAction(parseActionSellNFT),
    fetchDataSellNFT,
    formatSecurityEngineSellNFT,
  ],
  [parseAction, fetchActionRequiredData, formatSecurityEngineContext],
])('SellNFT', async (_parseAction, _fetchData, _format) => {
  const actionData = _parseAction({
    type: 'typed_data',
    data: parseTxData['action'],
    typedData: txData,
    sender: SENDER,
  });
  expect(actionData).toMatchSnapshot('parseActionSellNFT');

  const requireData = await _fetchData({
    type: 'typed_data',
    actionData,
    chainId: ETH_CHAIN_ID,
    walletProvider,
    apiProvider,
    sender: SENDER,
  });
  expect(requireData).toMatchSnapshot('fetchDataSellNFT');

  const ctx = await _format({
    type: 'typed_data',
    actionData,
    requireData,
    chainId: ETH_CHAIN_ID,
    isTestnet: false,
    provider: formatProvider,
    origin: ORIGIN,
  });
  expect(ctx).toMatchSnapshot('formatSecurityEngineSellNFT');
});
