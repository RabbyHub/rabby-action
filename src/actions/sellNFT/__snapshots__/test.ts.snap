// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SellNFT: fetchDataSellNFT 1`] = `
{
  "bornAt": 1591388241,
  "extraState": {
    "hasInteraction": [Function],
    "receiverHasTransfer": [Function],
  },
  "hasInteraction": true,
  "id": "0xeaf5453b329eb38be159a872a6ce91c9a8fb0260",
  "protocol": {
    "id": "uniswap2",
    "logo_url": "https://static.debank.com/image/project/logo_url/uniswap2/87a541b3b83b041c8d12119e5a0d19f0.png",
    "name": "Uniswap V2",
  },
  "rank": 65,
  "receiverInWallet": false,
  "sender": "0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85",
  "unexpectedAddr": null,
}
`;

exports[`SellNFT: fetchDataSellNFT 2`] = `
{
  "bornAt": 1591388241,
  "extraState": {
    "hasInteraction": [Function],
    "receiverHasTransfer": [Function],
  },
  "hasInteraction": true,
  "id": "0xeaf5453b329eb38be159a872a6ce91c9a8fb0260",
  "protocol": {
    "id": "uniswap2",
    "logo_url": "https://static.debank.com/image/project/logo_url/uniswap2/87a541b3b83b041c8d12119e5a0d19f0.png",
    "name": "Uniswap V2",
  },
  "rank": 65,
  "receiverInWallet": false,
  "sender": "0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85",
  "unexpectedAddr": null,
}
`;

exports[`SellNFT: formatSecurityEngineSellNFT 1`] = `
{
  "sellNFT": {
    "chainId": "eth",
    "from": "0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85",
    "id": "0xeaf5453b329eb38be159a872a6ce91c9a8fb0260",
    "receiveTokenIsFake": false,
    "receiveTokenIsScam": false,
    "receiver": "0xfac07fa759cae79cf85fa4f7c947ace0a2d5ed8c",
    "specificBuyer": undefined,
  },
}
`;

exports[`SellNFT: formatSecurityEngineSellNFT 2`] = `
{
  "sellNFT": {
    "chainId": "eth",
    "from": "0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85",
    "id": "0xeaf5453b329eb38be159a872a6ce91c9a8fb0260",
    "receiveTokenIsFake": false,
    "receiveTokenIsScam": false,
    "receiver": "0xfac07fa759cae79cf85fa4f7c947ace0a2d5ed8c",
    "specificBuyer": undefined,
  },
}
`;

exports[`SellNFT: parseActionSellNFT 1`] = `
{
  "actionType": "sell_nft_order",
  "brand": undefined,
  "chainId": 137,
  "contractId": "0xeaf5453b329eb38be159a872a6ce91c9a8fb0260",
  "sellNFT": {
    "expire_at": 4.523128485832664e+74,
    "pay_nft": {
      "amount": 90,
      "attributes": [],
      "chain": "matic",
      "collection": {
        "chain": "matic",
        "credit_score": null,
        "description": null,
        "floor_price": 0,
        "id": "matic:0xf494d8c3a421b6175cb1d89ea70715e1932637a9",
        "is_core": false,
        "is_scam": true,
        "is_suspicious": false,
        "is_verified": true,
        "logo_url": "",
        "name": "PLANT.z - BLOOMER",
      },
      "collection_id": "matic:0xf494d8c3a421b6175cb1d89ea70715e1932637a9",
      "content": "",
      "content_type": null,
      "contract_id": "0xf494d8c3a421b6175cb1d89ea70715e1932637a9",
      "description": null,
      "detail_url": "https://opensea.io/assets/0xf494d8c3a421b6175cb1d89ea70715e1932637a9/1",
      "id": "2374a8d8f35ed510c74ad7838f06a322",
      "inner_id": "1",
      "is_erc1155": true,
      "is_erc721": false,
      "name": "#1",
      "pay_token": null,
      "symbol": "PLANT",
      "thumbnail_url": "",
      "total_supply": 420,
    },
    "receive_token": {
      "amount": 41.454,
      "chain": "matic",
      "credit_score": 100000000,
      "decimals": 18,
      "display_symbol": null,
      "id": "matic",
      "is_core": true,
      "is_scam": false,
      "is_suspicious": false,
      "is_verified": true,
      "is_wallet": true,
      "logo_url": "https://static.debank.com/image/matic_token/logo_url/matic/6f5a6b6f0732a7a235131bd7804d357c.png",
      "name": "MATIC",
      "optimized_symbol": "MATIC",
      "price": 0.525,
      "price_24h_change": -0.06766116142781037,
      "protocol_id": "",
      "raw_amount": "41454000000000000000",
      "symbol": "MATIC",
      "time_at": null,
    },
    "receiver": "0xfac07fa759cae79cf85fa4f7c947ace0a2d5ed8c",
    "takers": [],
  },
  "sender": "0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85",
}
`;

exports[`SellNFT: parseActionSellNFT 2`] = `
{
  "actionType": "sell_nft_order",
  "brand": undefined,
  "chainId": 137,
  "contractId": "0xeaf5453b329eb38be159a872a6ce91c9a8fb0260",
  "sellNFT": {
    "expire_at": 4.523128485832664e+74,
    "pay_nft": {
      "amount": 90,
      "attributes": [],
      "chain": "matic",
      "collection": {
        "chain": "matic",
        "credit_score": null,
        "description": null,
        "floor_price": 0,
        "id": "matic:0xf494d8c3a421b6175cb1d89ea70715e1932637a9",
        "is_core": false,
        "is_scam": true,
        "is_suspicious": false,
        "is_verified": true,
        "logo_url": "",
        "name": "PLANT.z - BLOOMER",
      },
      "collection_id": "matic:0xf494d8c3a421b6175cb1d89ea70715e1932637a9",
      "content": "",
      "content_type": null,
      "contract_id": "0xf494d8c3a421b6175cb1d89ea70715e1932637a9",
      "description": null,
      "detail_url": "https://opensea.io/assets/0xf494d8c3a421b6175cb1d89ea70715e1932637a9/1",
      "id": "2374a8d8f35ed510c74ad7838f06a322",
      "inner_id": "1",
      "is_erc1155": true,
      "is_erc721": false,
      "name": "#1",
      "pay_token": null,
      "symbol": "PLANT",
      "thumbnail_url": "",
      "total_supply": 420,
    },
    "receive_token": {
      "amount": 41.454,
      "chain": "matic",
      "credit_score": 100000000,
      "decimals": 18,
      "display_symbol": null,
      "id": "matic",
      "is_core": true,
      "is_scam": false,
      "is_suspicious": false,
      "is_verified": true,
      "is_wallet": true,
      "logo_url": "https://static.debank.com/image/matic_token/logo_url/matic/6f5a6b6f0732a7a235131bd7804d357c.png",
      "name": "MATIC",
      "optimized_symbol": "MATIC",
      "price": 0.525,
      "price_24h_change": -0.06766116142781037,
      "protocol_id": "",
      "raw_amount": "41454000000000000000",
      "symbol": "MATIC",
      "time_at": null,
    },
    "receiver": "0xfac07fa759cae79cf85fa4f7c947ace0a2d5ed8c",
    "takers": [],
  },
  "sender": "0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85",
}
`;
